packages/editor/src/package/storage-format.ts

```diff
@@ -0,0 +1,133 @@
+
+type Migration = (state: { version: number }) => { version: number }
+
+const migrations: Migration[] = [
+  // Migration: Add `editorVersion`
+  (state): StorageFormat => {
+    const expectedType = t.type({
+      type: t.literal(documentType),
+      variant: EditorVariantType,
+      version: t.number,
+      dateModified: t.string,
+      document: t.type({
+        plugin: t.string,
+        state: t.unknown,
+      }),
+    })
+    if (!expectedType.is(state))
+      throw new Error(
+        `Unexpected type during migration. Expected ${JSON.stringify(expectedType)} but got ${JSON.stringify(state)}`
+      )
+
+    const editorVersion = getEditorVersion()
+
+    return {
+      ...state,
+      editorVersion,
+    }
+  },
+  // ...
+  // Add new migrations here. Make sure the last one returns the new StorageFormat.
+]
+
+const currentVersion = migrations.length
+
+export function createEmptyDocument(
+  editorVariant: EditorVariant
+): StorageFormat {
+  return {
+    type: documentType,
+    variant: editorVariant,
+    version: currentVersion,
+    editorVersion: getEditorVersion(),
+    dateModified: getCurrentDatetime(),
+    document: {
+      plugin: 'type-generic-content',
+      state: {
+        content: {
+          plugin: 'rows',
+          state: [
+            {
+              plugin: 'text',
+              state: [
+                {
+                  type: 'p',
+                  children: [{ text: '' }],
+                },
+              ],
+            },
+          ],
+        },
+      },
+    },
+  }
+}
+
+export function migrate(stateBeforeMigration: unknown): {
+  migratedState: StorageFormat
+  stateChanged: boolean
+} {
+  if (!t.type({ version: t.number }).is(stateBeforeMigration))
+    throw new Error(
+      `Missing property 'version: number' in state while trying to perform migrations. Got ${JSON.stringify(stateBeforeMigration)}`
+    )
+
+  // Only apply migrations for this version number and higher
+  const nextMigrationIndex = stateBeforeMigration.version
+
+  // Create deep copy
+  let migratedState = JSON.parse(JSON.stringify(stateBeforeMigration)) as {
+    version: number
+  }
+  let stateChanged = false
+  for (let i = nextMigrationIndex; i < migrations.length; i++) {
+    migratedState = migrations[i](migratedState)
+    stateChanged = true
+    migratedState.version = i + 1
+  }
+
+  if (!StorageFormatType.is(migratedState))
+    throw new Error(
+      'Storage format after migrations does not match StorageFormatType'
+    )
+
+  if (stateChanged) {
+    migratedState.editorVersion = getEditorVersion()
+    migratedState.dateModified = getCurrentDatetime()
+  }
+
+  return { migratedState, stateChanged }
+}
+
+/** The variant of the Serlo editor that created this saved data */
+const EditorVariantType = t.union([
+  t.literal('https://github.com/serlo/serlo-editor-for-edusharing'),
+  t.literal('https://github.com/serlo/serlo-editor-as-lti-tool/'),
```

Comment:
wouldn't the serlo.org editor also be a variant?

Reply:
(ah I remember you telling me this is only inside the package right now)

Reply:
Yes, hopefully soon we can add `serlo.org` to this list. :)